### ISSUE TITLE
Example of passing an image to Tensorflow model?

### ISSUE THREAD

**Issue by bzburr** [Microsoft/vs-tools-for-ai#25]
HI there

are there any c# examples of how to correctly pass a image across to a tensorflow model in ai tools after the Model project has been made?

cheers

**Comment by shishaochen** [comment]
For static image, please refer to the [Handwriting recognition](https://github.com/Microsoft/samples-for-ai/blob/master/projects/MnistDemo/MNIST.App/MainWindow.cs#L65) in our [sample repository](https://github.com/Microsoft/samples-for-ai).

**Comment by bzburr** [comment]
hmm weird

the Model Inference project created a function called "Serving_default"
When i add the new project into my main Solution this Serving_default function fails and wont let me step into it. However if I copy the code to a new function,...i can then step into my new function  - is this by design?

Also Modelmanager RunModel throws an exception "System.Runtime.InteropServices.SEHException: 'External component has thrown an exception.'"

Where does one find more info about why its complaining?

**Comment by shishaochen** [comment]
@bzburr The external component normally means the TensorFlow C++ kernel which is included in the Microsoft.ML.Scoring NuGet package.
The error causes vary a lot like data shape mismatch, x64 compiling option disabled, etc.
Could you please share your model and sample input if no privacy issue?

**Comment by bzburr** [comment]
Happy to :)
```

My calling code....have i incorrectly built the enumerable?

 public int ClassifyImage(string imageFile)
        {

            System.Drawing.Image loadedImage = System.Drawing.Image.FromFile(imageFile);
            Bitmap clonedBmp = new Bitmap(ImageSize, ImageSize);
            Graphics gNormalized = Graphics.FromImage(clonedBmp);
            gNormalized.DrawImage(loadedImage, 0, 0, ImageSize, ImageSize);
            clonedBmp.Save(@"c:\temp\annoy.jpg", ImageFormat.Jpeg);

            var image = new List<float>(ImageSize * ImageSize);
            for (var x = 0; x < ImageSize; x++)
            {
                for (var y = 0; y < ImageSize; y++)
                {
                    var color = clonedBmp.GetPixel(y, x);
                    // c. normalize value
                    //      RGB to gray
                    //      Range to [-0.5,0.5]
                    image.Add((float)(0.5 - (color.R + color.G + color.B) / (3.0 * 255)));
                }
            }

            AiToolsImageClassifier.ImageClassifier classifier = new AiToolsImageClassifier.ImageClassifier();
            List<IEnumerable<float>> vs = new List<IEnumerable<float>>();
            
            vs.Add(image);
            classifier.Infer(vs);  //for testing

            var res = classifier.Serving_default(vs);
            .....


        }
```
[ public IEnumerable<IEnumerable<float>> Serving_default(IEnumerable<IEnumerable<float>> imageBatch)
        {
            List<float> imageCombined = new List<float>();
            foreach (var input in imageBatch)
            {
                imageCombined.AddRange(input);
            }

            List<Tensor> result = manager.RunModel(
                modelName,
                int.MaxValue,
                serving_defaultInputNames,
                new List<Tensor> { new Tensor(imageCombined, new List<long> { imageBatch.LongCount(), 299, 299, 3 }) },
                serving_defaultOutputNames
            );

            int predictionBatchNum = (int)result[0].GetShape()[0];
            int predictionBatchSize = (int)result[0].GetShape().Aggregate((a, x) => a * x) / predictionBatchNum;
            for (int batchNum = 0, offset = 0; batchNum < predictionBatchNum; batchNum++, offset += predictionBatchSize)
            {
                List<float> tmp = new List<float>();
                result[0].CopyTo(tmp, offset, predictionBatchSize);
                yield return tmp;
            }
        }

//for testing...
        public void Infer(IEnumerable<IEnumerable<float>> imageBatch)
        {
            List<float> imageCombined = new List<float>();
            foreach (var input in imageBatch)
            {
                imageCombined.AddRange(input);
            }

            List<Tensor> result = manager.RunModel(
                modelName,
                int.MaxValue,
                serving_defaultInputNames,
                new List<Tensor> { new Tensor(imageCombined, new List<long> { imageBatch.LongCount(), 299, 299, 3 }) },
                serving_defaultOutputNames
            );

            int predictionBatchNum = (int)result[0].GetShape()[0];
            int predictionBatchSize = (int)result[0].GetShape().Aggregate((a, x) => a * x) / predictionBatchNum;
            for (int batchNum = 0, offset = 0; batchNum < predictionBatchNum; batchNum++, offset += predictionBatchSize)
            {
                List<float> tmp = new List<float>();
                result[0].CopyTo(tmp, offset, predictionBatchSize);
                //yield return ;
                return;
                
            }
        }
](url)

**Comment by linmajia** [comment]
@bzburr , thank you very much for the feedback.

if you cannot step into the "Serving_default" function in the inference project, you may take a look at the following things:
1. You could launch the "Modules" window (Debug -> Window -> Modules) during debugging to check whether the symbol file (PDB) is loaded correctly.
2. You may have "Just My Code" enabled. Try disabling it from "Tools -> Options -> Debugger -> Just My Code".

**Comment by bzburr** [comment]
So Debug -> Window -> Modules says symbols loaded for the Model Interface 
Just my code was disabled ...and i'm compiling all projects as 0x64

**Comment by shishaochen** [comment]
@bzburr Could you send your files of model and image to my email? You can find it in my GitHub profile page.

**Comment by shishaochen** [comment]
@bzburr The model you sent is the processed SavedModel file without variable files. I cannot restore it.
Could you please share the origin one with complete graph, variables?

**Comment by bzburr** [comment]
its a savedmodel from Microsofts - customvision.ai - i created it there and exported it.
thats all it came with.

I i  can run retrain.py and so a savemodel from that if you'd like....

**Comment by shishaochen** [comment]
@bzburr CustomVision model should be OK as I tested before.
I don't need a SavedModel format, the original format like frozen graph and checkpoint is OK for me.

**Comment by shishaochen** [comment]
@bzburr Thanks for your sharing:)
It seems that you were doing transfer learning based on Inception model from TensorFlow Hub.
Let me investigate more before give your response.

**Comment by shishaochen** [comment]
@bzburr I find your error!!!
As the generated C# file told, the shape of the input tensor should be (299, 299, 3).
![image](https://user-images.githubusercontent.com/17514649/43364048-0fb4de40-9345-11e8-80d7-2527ac851a4e.png)
But in your code, you only provided an array with size (299, 299) which led to the Tensor construction error. You can find the error from the VS debug window.
![image](https://user-images.githubusercontent.com/17514649/43364110-83864d62-9346-11e8-806e-aeb232f2486a.png)
The largest BUG of your program is **you simply copying the image processing code I shared without any change**. My shared sample code works for MNIST dataset but your scenario.
Please parse your image with correct processing steps according to your training code.

**Comment by bzburr** [comment]
That cant be right ...you said to follow the NMIST example

https://github.com/Microsoft/samples-for-ai/blob/master/projects/MnistDemo/MNIST.App/MainWindow.cs#L65




the code you quote 

https://user-images.githubusercontent.com/17514649/43364048-0fb4de40-9345-11e8-80d7-2527ac851a4e.png

is from the Auto generated code is pretty much a copy of the NMIST stuff.

NMISTS Infer function looks like 
[
public IEnumerable<IEnumerable<long>> Infer(IEnumerable<IEnumerable<float>> inputsBatch)
        {
            List<float> inputsCombined = new List<float>();
            foreach (var input in inputsBatch)
            {
                inputsCombined.AddRange(input);
            }

            List<Tensor> result = manager.RunModel(
                modelName,
                int.MaxValue,
                inferInputNames,
                new List<Tensor> { new Tensor(inputsCombined, new List<long> { inputsBatch.LongCount(), 28, 28, 1 }) },
                inferOutputNames
            );

            int outputsBatchNum = (int)result[0].GetShape()[0];
            int outputsBatchSize = (int)result[0].GetShape().Aggregate((a, x) => a * x) / outputsBatchNum;
            for (int batchNum = 0, offset = 0; batchNum < outputsBatchNum; batchNum++, offset += outputsBatchSize)
            {
                List<long> tmp = new List<long>();
                result[0].CopyTo(tmp, offset, outputsBatchSize);
                yield return tmp;
            }
        }](url)


The auto generated code is the same for both.
Now, over in MNST we see

[private void writeArea_MouseUp(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                /*
                  * Overview : normalize the picture and send it to the model
                  * The model can infer multiple images at once and return multiple results
                  * The input of mnist model is a list of "image" 
                  * Each "image" is a float list which records every pixel's grayscale value of a 28*28 image
                  * 
                  * In this demo program, we only send one image to infer at once
                  * So, before the inference, we need to resize the image, covert it to gray and get the grayscale value of each pixel    
                  * Then we need to put the image into a new list and send the list to the model
                  * When displaying the result, we need to get the predicted digit of the highest probability of the first image
                */

                // 1. begin to normalize data
                Bitmap clonedBmp = new Bitmap(MnistImageSize, MnistImageSize);
                Graphics gNormalized = Graphics.FromImage(clonedBmp);

                // a. normalize the size to 28*28 as the training input
                gNormalized.DrawImage(writeArea.Image, 0, 0, MnistImageSize, MnistImageSize);

                // b. normalize the data structure to a float list as training input
                var image = new List<float>(MnistImageSize * MnistImageSize);
                for (var x = 0; x < MnistImageSize; x++)
                {
                    for (var y = 0; y < MnistImageSize; y++)
                    {
                        var color = clonedBmp.GetPixel(y, x);
                        // c. normalize value
                        //      RGB to gray
                        //      Range to [-0.5,0.5]
                        image.Add((float)(0.5 - (color.R + color.G + color.B) / (3.0 * 255)));
                    }
                }
                // end to normalize data

                // 2. inferencing
                //      a. put the image into a new List 
                //      b. get the predicted digit of the highest probability of the first image
                outputText.Text = (model.Infer(new List<IEnumerable<float>> { image })).First().First().ToString();
            }
        }](url)


I don't see anywhere in NMIST thats any difference between what I'm doing .

Can you show me exactly where my stuff differs from the NMIST  because i'm not seeing it :(

**Comment by bzburr** [comment]
ok so 
 - i need to 

```
for (var x = 0; x < ImageSize; x++)
                {
                    for (var y = 0; y < ImageSize; y++)
                    {
                        var color = clonedBmp.GetPixel(y, x);

                        image.Add((float)(  (color.R  ) / ( 255)));
                        image.Add((float)(  ( color.G ) / (  255)));
                        image.Add((float)(  (color.B) / (  255)));
                    }
                }
```

The NMIST doesn't do that it just 
```

for (var x = 0; x < ImageSize; x++)
                {
                    for (var y = 0; y < ImageSize; y++)
                    {
                        var color = clonedBmp.GetPixel(y, x);

                        // c. normalize value
                        //      RGB to gray
                        //      Range to [-0.5,0.5]
                         image.Add((float)(0.5 - (color.R + color.G + color.B) / (3.0 * 255)));
                    }
                }
```

and given NMIST Infer code wants a 

List<Tensor> result = manager.RunModel(
                modelName,
                int.MaxValue,
                inferInputNames,
                new List<Tensor> { new Tensor(inputsCombined, new List<long> { inputsBatch.LongCount(), 28, 28, **1** }) },
                inferOutputNames
            );

ahh thanks :)

**Comment by shishaochen** [comment]
@bzburr Different models have different image processing steps.
For MNIST sample, it need convert a image to a black-white array and normalize their gray degrees.
For the model you use, you need investigate the original training script to find out how it convert a raw image into a 299*299*3 array.

**Comment by bzburr** [comment]
var image = new List<float>(ImageSize * ImageSize *3);
                for (var x = 0; x < ImageSize; x++)
                {
                    for (var y = 0; y < ImageSize; y++)
                    {
                        var color = clonedBmp.GetPixel(y, x);

                        image.Add((float)(   (color.R  ) / ( 255)));
                        image.Add((float)(   ( color.G ) / (  255)));
                        image.Add((float)(   (color.B) / (  255)));
                  
                    }
                }


how does that look to you?

**Comment by shishaochen** [comment]
@bzburr Although your change can make shape right and the code pass, the value you provide may not be expected by your TensorFlow model which can bring a wrong result.
